### PR TITLE
Provide span for declaration of captured variables

### PR DESCRIPTION
**Pull request by davidtwco**
Part of #52663.

r? @nikomatsakis

### CODE REVIEW

File: src/librustc_mir/borrow_check/move_errors.rs [hunk 1]
```diff
@@ -237,17 +259,11 @@ impl<'a, 'gcx, 'tcx> MirBorrowckCtxt<'a, 'gcx, 'tcx> {
...
-                                            Place::Projection(ref proj) => {
-                                                proj.base == Place::Local(Local::new(1))
-                                            }
-                                            Place::Promoted(_) |
-                                            Place::Local(_) | Place::Static(_) => unreachable!(),
-                                        }
-                                    } =>
-                            {
+                                if !self.mir.upvar_decls.is_empty() &&
+                                    original_path.strip_deref_projections()
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/move_errors.rs]: hmm, is `strip_deref_projections` enough?

Try a case like this:

```rust
{
    let x = (vec![22], vec![44]);
    expect_fn(|| drop(x.0));
}
```

does that work?

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: src/librustc_mir/borrow_check/move_errors.rs (hunk 1 above)
Comment (davidtwco) [reply src/librustc_mir/borrow_check/move_errors.rs]: Fixed.

File: src/test/ui/nll/issue-52663-span-decl-captured-variable.stderr [hunk 2]
```diff
@@ -0,0 +1,9 @@
+error[E0507]: cannot move out of captured variable in an `Fn` closure
+  --> $DIR/issue-52663-span-decl-captured-variable.rs:20:26
+   |
+LL |        expect_fn(|| drop(x.0));
```
Comment (nikomatsakis) [inline src/test/ui/nll/issue-52663-span-decl-captured-variable.stderr]: shouldn't we highlight something here?

Comment (davidtwco) [reply src/test/ui/nll/issue-52663-span-decl-captured-variable.stderr]: Oops, forgot to update the label logic with the same changes that made that branch get used in the first place for complex cases like this one. Fixed now.

**Review by nikomatsakis — APPROVED** [review]

**Comment by davidtwco** [comment]
Oops, fixed!

**Comment by kennytm** [comment]
@bors retry #40474

```
[01:59:15]  Downloading idna v0.1.0
[01:59:46] error: unable to get packages from source
[01:59:46] 
[01:59:46] Caused by:
[01:59:46]   [18] Transferred a partial file (transfer closed with 9418 bytes remaining to read)
[01:59:46] thread 'main' panicked at 'tests failed for https://github.com/iron/iron', tools/cargotest/main.rs:100:9
[01:59:46] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:59:46] 
```

**Comment by kennytm** [comment]
@bors r-

Merge conflict.

**Comment by davidtwco** [comment]
Rebased.

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**